### ISSUE TITLE
no sound with wdm/ks under win8.1

### ISSUE THREAD

**Issue by piotx** [spatialaudio/python-sounddevice#36]
I've compiled portaudio myself (with Visual C++ 2008 Express Edition), and tried out binaries from https://github.com/adfernandes/precompiled-portaudio-windows but there is no sound on win 8.1, when using wdm/ks...

any ideas how to get around it?

**Comment by mgeier** [comment]
Did you try to use the bundled DLL?

**Comment by piotx** [comment]
yes. wdm-ks isn't available then: default_output_device: -1,
default_input_device: -1

but I also get some problems with utf8, when querying for devices or trying
to use wasapi:

C:\snd_scripts_f>python -m sounddevice
Traceback (most recent call last):
  File "C:\Python27\lib\runpy.py", line 174, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "C:\Python27\lib\runpy.py", line 72, in _run_code
    exec code in run_globals
  File "C:\Python27\lib\site-packages\sounddevice.py", line 2517, in
<module>
    print(query_devices())
  File "C:\Python27\lib\site-packages\sounddevice.py", line 613, in
query_device
s
    for i in range(_check(_lib.Pa_GetDeviceCount())))
  File "C:\Python27\lib\site-packages\sounddevice.py", line 613, in
<genexpr>
    for i in range(_check(_lib.Pa_GetDeviceCount())))
  File "C:\Python27\lib\site-packages\sounddevice.py", line 622, in
query_device
s
    _lib.Pa_HostApiTypeIdToHostApiIndex(_lib.paMME)) else 'utf-8'),
  File "C:\Python27\lib\encodings\utf_8.py", line 16, in decode
    return codecs.utf_8_decode(input, errors, True)
UnicodeDecodeError: 'utf8' codec can't decode byte 0xe4 in position 48:
invalid
continuation byte

2016-08-16 19:42 GMT+02:00 Matthias Geier <notifications@github.com>:

**Comment by piotx** [comment]
just to let you know; wdm ks works perfect on windows 7; sound latency (measured with an oscilloscope) is between 0.5 and 2 ms; that's very good, thanks a lot!

**Comment by piotx** [comment]
I think the issue can be closed; I can't reproduce the problem on win8 anymore. wdm-ks worked today, when I specified the location of the audioport.dll with 

    _lib = _ffi.dlopen('c:\python27\Lib\site-packages\_sounddevice_data\libportaudio32bit.dll' 
Maybe the portaudio.dll of pyo was loaded, without my knowing

**Comment by mgeier** [comment]
OK, that's possible. If `dlopen('portaudio')` succeeds, the bundled DLL is not used.

I guess this is a feature?
This is necessary for Linux, I don't know if I can/should disable this for Windows ...?

You can try to check if the PortAudio versions are different:

```python
import sounddevice as sd
sd.get_portaudio_version()
```

You might get some information about the used library with this:

```python
from ctypes.util import find_library
find_library('portaudio')
```

Just out of curiosity, where is pyo's DLL installed?

**Comment by piotx** [comment]
thanks for the infos. It's actually a nice thing, that one can choose the
portaudio.dll. But it can mislead...

pyo's dll is installed in the main site-package folder:
pythonpath/Lib/site-packages/portaudio.dll

2016-08-25 14:25 GMT+02:00 Matthias Geier <notifications@github.com>:

**Comment by piotx** [comment]
that's ok now.
now, 
I get however an error, if I call 
```python
sd.play(data, device=9, extra_settings=wasapi_exclusive)```

c:\python27\lib\site-packages\sounddevice-0.3.4-py2.7.egg\sounddevice.py in __in
it__(self, kind, samplerate, blocksize, device, channels, dtype, latency, extra_
settings, callback_wrapper, finished_callback, clip_off, dither_off, never_drop_
input, prime_output_buffers_using_stream_callback)
    908                                   samplerate, blocksize, stream_flags,
    909                                   self._callback, _ffi.NULL),
--> 910                'Error opening {0}'.format(self.__class__.__name__))
    911
    912         # dereference PaStream** --> PaStream*

c:\python27\lib\site-packages\sounddevice-0.3.4-py2.7.egg\sounddevice.py in _che
ck(err, msg)
   2601         else:
   2602             msg += _ffi.string(_lib.Pa_GetErrorText(err)).decode()
-> 2603         raise PortAudioError(msg)
   2604     return err
   2605

PortAudioError: Error opening OutputStream: Incompatible host API specific strea
m info

`sd.query_devices()` shows:
 9 Speakers (Realtek High Definition Audio), Windows WASAPI (0 in, 2 out)

there seems to be another small issue with `sd.default.extra_settings = wasapi_exclusive`;

I get the following error message, if I run it:
c:\python27\lib\site-packages\sounddevice-0.3.4-py2.7.egg\sounddevice.py in __i
it__(self, kind, samplerate, blocksize, device, channels, dtype, latency, extra
settings, callback_wrapper, finished_callback, clip_off, dither_off, never_drop
input, prime_output_buffers_using_stream_callback)
    887             parameters, self._dtype, self._samplesize, samplerate = \
    888                 _get_stream_parameters(kind, device, channels, dtype, l
tency,
--> 889                                        extra_settings, samplerate)
    890             self._device = parameters.device
    891             self._channels = parameters.channelCount

c:\python27\lib\site-packages\sounddevice-0.3.4-py2.7.egg\sounddevice.py in _ge
_stream_parameters(kind, device, channels, dtype, latency, extra_settings, samp
erate)
   2523         latency = default.latency[kind]
   2524     if extra_settings is None:
-> 2525         extra_settings = default.extra_settings[kind]
   2526     if samplerate is None:
   2527         samplerate = default.samplerate

TypeError: 'WasapiSettings' object has no attribute '__getitem__'

**Comment by mgeier** [comment]
Thanks for testing! Just for the record, this should have been an answer to https://github.com/spatialaudio/python-sounddevice/issues/35#issuecomment-242708385.

I fixed the second problem in a new commit to #39, can you please try again?

I don't know what's causing the first error, probably it's related to the second one ...?

**Comment by mgeier** [comment]
I'm closing this issue since it was caused by pyo's `portaudio.dll`.
